src/server/service/debug.rs

```diff
@@ -0,0 +1,203 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::error;
+use std::boxed::FnBox;
+use std::sync::Arc;
+use std::fmt::{self, Display};
+
+use grpc::{RpcContext, RpcStatus, RpcStatusCode, ServerStreamingSink, UnarySink};
+use futures::Future;
+use kvproto::debugpb_grpc;
+use kvproto::debugpb::*;
+use rocksdb::DB;
+
+use util::worker::{Runnable, Scheduler};
+use raftstore::store::keys;
+use raftstore::store::engine::Peekable;
+use storage::{CF_DEFAULT, CF_LOCK, CF_RAFT, CF_WRITE};
+
+use super::make_callback;
+
+pub type Callback = Box<FnBox(Result<Response, Error>) + Send>;
+
+#[derive(Clone)]
+pub struct Service {
+    scheduler: Scheduler<Request>,
+}
+
+impl Service {
+    pub fn new(scheduler: Scheduler<Request>) -> Service {
+        Service { scheduler }
+    }
+}
+
+impl debugpb_grpc::Debug for Service {
+    fn get(&self, ctx: RpcContext, mut req: GetRequest, sink: UnarySink<GetResponse>) {
+        const LABEL: &'static str = "debug_get";
+
+        let (cb, future) = make_callback();
+        let req = Request::Get {
+            cf: req.get_cf(),
+            key_encoded: req.take_key_encoded(),
+            callback: cb,
```

Comment:
Does gRPC support pipeline on client side? If there are multi requests on same session, here we could raise the second response before the first.

Reply:
After test, it seems like gRPC has already deal with async client pipeline internally.
Any **wait** on `ClientUnaryReceiver` will keep blocking until all pending async get are resolved.
So, we can ignore this.